File: src/main/java/io/protobase/f7/formulas/math/QUOTIENT.java
package io.protobase.f7.formulas.math;

import io.protobase.f7.errors.DivException;
import io.protobase.f7.errors.F7Exception;
import io.protobase.f7.formulas.AbstractFormula;
import io.protobase.f7.formulas.FormulaName;
import io.protobase.f7.models.GridColumnRowKey;
import io.protobase.f7.utils.Converters;

import java.util.function.BiFunction;
import java.util.function.Function;

public class QUOTIENT extends AbstractFormula {
  public static FormulaName NAME = FormulaName.QUOTIENT;
  public static QUOTIENT SELF = new QUOTIENT();

  public QUOTIENT() {
    super();
  }

  public QUOTIENT(Function<Object, Object> lookup, BiFunction<GridColumnRowKey, Object, Object> collateralLookup) {
    super(lookup, collateralLookup);
  }

  @Override
  public Object internal(GridColumnRowKey origin, Object... values) {
    checkLength(values.length, 2, NAME);
    Object first = Converters.first(collateralLookup.apply(origin, values[0]));
    Object second = Converters.first(collateralLookup.apply(origin, values[1]));
    if (first instanceof F7Exception) {
      return first;
    }
    if (second instanceof F7Exception) {
      return second;
    }
    double dividend = Converters.toDouble(first);
    double divisor = Converters.toDouble(second);
    if (divisor == 0 || divisor == -0) {
      throw new DivException("Formula DIVIDE parameter 2 cannot be zero.");
    }
    return ROUNDDOWN.SELF.apply(null, dividend / divisor);
  }
}

File: src/main/js/formulas/math/QUOTIENT.ts
import { DivException } from "../../errors/DivException";
import { F7Exception } from "../../errors/F7Exception";
import { SheetColumnRowKey } from "../../models/common/SheetColumnRowKey";
import { Complex } from "../../models/common/Types";
import { Converters } from "../../utils/Converters";
import { Numbers } from "../../utils/Numbers";
import { AbstractFormula } from "../AbstractFormula";
import { FormulaName } from "../FormulaName";
import { ROUNDDOWN } from "./ROUNDDOWN";

export class QUOTIENT extends AbstractFormula {
  static SELF: QUOTIENT = new QUOTIENT();
  NAME = FormulaName.QUOTIENT;

  internal(origin: SheetColumnRowKey, ...values: Array<Complex>) {
    AbstractFormula.checkLength(values.length, 2, this.NAME);
    const first = Converters.first(this.collateralLookup(origin, values[0]));
    const second = Converters.first(this.collateralLookup(origin, values[1]));
    if (first instanceof F7Exception) {
      return first;
    }
    if (second instanceof F7Exception) {
      return second;
    }
    const dividend = Converters.toNumber(first);
    const divisor = Converters.toNumber(second);
    if (Numbers.isZero(divisor)) {
      throw new DivException("Formula DIVIDE parameter 2 cannot be zero.");
    }
    return ROUNDDOWN.SELF.run(null, dividend / divisor);
  }
}
